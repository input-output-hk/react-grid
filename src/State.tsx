import React, { createContext, useState, FunctionComponent, useCallback, useEffect, ReactNode } from 'react'
import PropTypes from 'prop-types'

interface ScreenSizes {
  sm: number,
  md: number,
  lg: number,
  xl: number
}

enum ScreenSize {
  xs = 'xs',
  sm = 'sm',
  md = 'md',
  lg = 'lg',
  xl = 'xl'
}

interface PartialScreenSizes {
  sm?: number,
  md?: number,
  lg?: number,
  xl?: number
}

interface DefaultState {
  screenSizes: ScreenSizes,
  screenSize: ScreenSize,
  updateScreenSizes?(newScreenSizes: PartialScreenSizes): void
}

interface StateProps {
  children: ReactNode,
  screenSizes?: PartialScreenSizes
}

const DEFAULT_SCREEN_SIZES: ScreenSizes = {
  sm: 576,
  md: 768,
  lg: 992,
  xl: 1200
}

const getScreenSize: Function = (screenSizes: ScreenSizes): ScreenSize => {
  const screenWidth: number = window.innerWidth || 0
  if (screenWidth < screenSizes.sm) return ScreenSize.xs
  if (screenWidth >= screenSizes.sm && screenWidth < screenSizes.md) return ScreenSize.sm
  if (screenWidth >= screenSizes.md && screenWidth < screenSizes.lg) return ScreenSize.md
  if (screenWidth >= screenSizes.lg && screenWidth < screenSizes.xl) return ScreenSize.lg
  if (screenWidth >= screenSizes.xl) return ScreenSize.xl
}

const Context: React.Context<DefaultState> = createContext({ screenSizes: DEFAULT_SCREEN_SIZES, screenSize: getScreenSize(DEFAULT_SCREEN_SIZES) })
const ContextProvider = Context.Provider
export const Consumer = Context.Consumer

export const Provider: FunctionComponent = (props: StateProps) => {
  const [ screenSizes, setScreenSizes ]: [ ScreenSizes, Function ] = useState({ ...DEFAULT_SCREEN_SIZES, ...props.screenSizes })
  const [ screenSize, setScreenSize ]: [ ScreenSize, Function ] = useState(getScreenSize(screenSizes))
  const updateScreenSizes = (newScreenSizes: PartialScreenSizes): void => setScreenSizes({ ...screenSizes, ...newScreenSizes })
  const onResize: EventListener = useCallback(() => setScreenSize(getScreenSize(screenSizes)), [])

  useEffect(() => {
    setScreenSize(getScreenSize(screenSizes))
    window.removeEventListener && window.removeEventListener('resize', onResize)
    window.addEventListener && window.addEventListener('resize', onResize)
    return () => window.removeEventListener && window.removeEventListener('resize', onResize)
  }, [ screenSizes ])

  return (
    <ContextProvider value={{ screenSizes, updateScreenSizes, screenSize }}>
      {props.children}
    </ContextProvider>
  )
}

Provider.propTypes = {
  children: PropTypes.node.isRequired,
  screenSizes: PropTypes.shape({
    xs: PropTypes.number,
    sm: PropTypes.number,
    md: PropTypes.number,
    lg: PropTypes.number,
    xl: PropTypes.number
  })
}
