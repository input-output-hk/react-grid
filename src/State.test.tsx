import React from 'react'
import ReactTestUtils from 'react-dom/test-utils'
import { mount } from 'enzyme'
import { Provider, Consumer } from './State'

const getWrapper = (props = {}) => {
  return mount((
    <Provider {...props}>
      <Consumer>
        {({ screenSizes, screenSize }) => (
          <div>
            <ul>
              {Object.keys(screenSizes).map(screenSize => (
                <li key={screenSize}>
                  {screenSizes[screenSize]}
                </li>
              ))}
            </ul>
            <span key='current screen size'>
              {screenSize}
            </span>
          </div>
        )}
      </Consumer>
    </Provider>
  ))
}

interface Global {
  triggerWindowEvent: Function,
  windowEventListeners: Object
}

declare const global: Global

describe('State', () => {
  afterEach(() => {
    global.windowEventListeners = {}
  })

  test('the default state is correct', () => {
    expect(getWrapper()).toMatchSnapshot()
  })

  describe('xs screens', () => {
    let wrapper
    beforeEach(() => {
      (window as any).innerWidth = 320
      wrapper = getWrapper()
    })

    test('the default state is correct', () => {
      expect(wrapper).toMatchSnapshot()
    })

    describe('when screen resized to sm screen', () => {
      beforeEach(() => {
        (window as any).innerWidth = 600
        ReactTestUtils.act(() => {
          global.triggerWindowEvent('resize')
        })
      })

      test('it updates the wrapper correctly', () => {
        wrapper.update()
        expect(wrapper).toMatchSnapshot()
      })
    })
  })

  describe('sm screens', () => {
    beforeEach(() => {
      (window as any).innerWidth = 600
    })

    test('the default state is correct', () => {
      expect(getWrapper()).toMatchSnapshot()
    })
  })

  describe('md screens', () => {
    beforeEach(() => {
      (window as any).innerWidth = 800
    })

    test('the default state is correct', () => {
      expect(getWrapper()).toMatchSnapshot()
    })
  })

  describe('lg screens', () => {
    beforeEach(() => {
      (window as any).innerWidth = 1100
    })

    test('the default state is correct', () => {
      expect(getWrapper()).toMatchSnapshot()
    })
  })

  describe('xl screens', () => {
    beforeEach(() => {
      (window as any).innerWidth = 1600
    })

    test('the default state is correct', () => {
      expect(getWrapper()).toMatchSnapshot()
    })
  })

  describe('when using custom break points', () => {
    let wrapper
    beforeEach(() => {
      wrapper = getWrapper({
        screenSizes: {
          sm: 200,
          md: 300,
          lg: 400,
          xl: 400
        }
      })
    })

    test('it renders correctly', () => {
      expect(wrapper).toMatchSnapshot()
    })
  })
})
