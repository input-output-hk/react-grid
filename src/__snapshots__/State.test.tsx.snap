// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`State lg screens the default state is correct 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      lg
    </span>
  </div>
</Provider>
`;

exports[`State md screens the default state is correct 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      md
    </span>
  </div>
</Provider>
`;

exports[`State sm screens the default state is correct 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      sm
    </span>
  </div>
</Provider>
`;

exports[`State the default state is correct 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      xl
    </span>
  </div>
</Provider>
`;

exports[`State when using custom break points it renders correctly 1`] = `
<Provider
  screenSizes={
    Object {
      "lg": 400,
      "md": 300,
      "sm": 200,
      "xl": 400,
    }
  }
>
  <div>
    <ul>
      <li
        key="sm"
      >
        200
      </li>
      <li
        key="md"
      >
        300
      </li>
      <li
        key="lg"
      >
        400
      </li>
      <li
        key="xl"
      >
        400
      </li>
    </ul>
    <span
      key="current screen size"
    >
      xl
    </span>
  </div>
</Provider>
`;

exports[`State xl screens the default state is correct 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      xl
    </span>
  </div>
</Provider>
`;

exports[`State xs screens the default state is correct 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      xs
    </span>
  </div>
</Provider>
`;

exports[`State xs screens when screen resized to sm screen it updates the wrapper correctly 1`] = `
<Provider>
  <div>
    <ul>
      <li
        key="sm"
      >
        576
      </li>
      <li
        key="md"
      >
        768
      </li>
      <li
        key="lg"
      >
        992
      </li>
      <li
        key="xl"
      >
        1200
      </li>
    </ul>
    <span
      key="current screen size"
    >
      sm
    </span>
  </div>
</Provider>
`;
